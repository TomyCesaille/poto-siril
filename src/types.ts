export type SpecFile = {
  name: string;
  path: string;
  fullPath: string;

  /**
   * File type: light, dark, bias, flat
   */
  type: string;
  /**
   * Bulb duration: 60.0s, 810.0ms
   */
  bulb: string;
  /**
   * Binning: Bin1, Bin2
   */
  bin: string;
  /**
   * Filter: L, R, G, B, H
   */
  filter: string;
  /**
   * Gain: 0, 100, 200, 300
   */
  gain: number;
  /**
   * Date and time: 20240707-002348
   */
  datetime: string;
  /**
   * Temperature: -10.0C, -9.9C
   */
  temperature: string;
  /**
   * Sequence number: 1, 60
   */
  sequence: number;
  /**
   * File extension: fit
   */
  extension: string;
};
