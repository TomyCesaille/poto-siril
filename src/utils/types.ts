type imageType = "Light" | "Dark" | "Bias" | "Flat";

export type ImageSpec = {
  /**
   * `Flat_520.0ms_Bin1_O_gain0` format.
   */
  setName: string;

  /**
   * File type: Light, Dark, Bias, Flat.
   */
  type: imageType;
  /**
   * Bulb duration: 60.0s, 810.0ms, etc...
   */
  bulb: string;
  /**
   * Binning: Bin1, Bin2, Bin 3 & Bin 4.
   */
  bin: string;
  /**
   * Filter: L, R, G, B, H... As Specified in the ASIAIR zwo wheel.
   * Optional, may not be present in the filename if no wheel.
   */
  filter: string | null;
  /**
   * Gain: 0, 100, 360, etc...
   */
  gain: number;
  /**
   * Date and time. `20240707-002348` format.
   */
};

export type FileImageSpec = ImageSpec & {
  /**
   * Name of the time in disk.
   */
  fileName: string;
  /**
   * File extension: `fit`.
   */
  extension: string;

  /**
   * Path to the ASIAIR dump directory.
   */
  sourceDirectory: string;
  /**
   * Path to the file in the ASIAIR dump directory.
   */
  sourceFilePath: string;

  /**
   * Path to the project directory.
   */
  projectDirectory: string;

  /**
   * Recorded time.
   */
  datetime: Date;

  /**
   * Sequence identifier. Is date of the first file of the sequence in format `20240626-010850`.
   */
  sequenceId: string;
  /**
   * Position in the sequence.
   */
  sequenceNumber: number;

  // TODO. Change to float Celcius.
  /**
   * Temperature: `-10.0C`, `-9.9C` format.
   */
  temperature: string;
};

/**
 * This represents a set of lights and its calibration frames.
 * Each set has unique filter + exposure + gain + bin combination.
 */
export type LayerSet = {
  filter: string;

  lightSet: string;
  lightsCount: number;

  flatSet: string;
  flatsCount: number;

  darkSet: string;
  darksCount: number;

  biasSet: string;
  biasesCount: number;

  lights: FileImageSpec[];
  flats: FileImageSpec[];
  darks: FileImageSpec[];
  biases: FileImageSpec[];
};

export type PotoProject = {
  generatedAt: Date;
  potoVersion: string;
  layerSets: LayerSet[];
};
