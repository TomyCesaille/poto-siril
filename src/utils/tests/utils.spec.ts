import { jest } from "@jest/globals"; // Import Jest globals

import fs from "fs";
import {
  getFileImageSpecFromFilename,
  getFitsFromDirectory,
  copyFileToProject,
  matchSetFile,
  getImageSpecFromSetName,
} from "../utils";
import { FileImageSpec, ImageSpec } from "../types";

describe("utils", () => {
  describe("getFileImageSpecFromFilename", () => {
    it("should match snapshot (ASIAIR plan target format)", () => {
      const file = new fs.Dirent();
      file.name =
        "Light_LDN 1093_120.0s_Bin1_H_gain100_20240707-002348_-10.0C_0001.fit";
      file.path = "input/bar";
      const projectDirectory = "project/bar";
      const previousFile = null;

      expect(
        getFileImageSpecFromFilename(file, projectDirectory, previousFile),
      ).toMatchSnapshot();
    });

    it("should match snapshot (filter named with spaces)", () => {
      const file = new fs.Dirent();
      file.name =
        "Light_LDN 1093_120.0s_Bin1_filter h_gain100_20240707-002348_-10.0C_0001.fit";
      file.path = "input/bar";
      const projectDirectory = "project/bar";
      const previousFile = null;

      const specs = getFileImageSpecFromFilename(
        file,
        projectDirectory,
        previousFile,
      );
      expect(specs).toMatchSnapshot();
      expect(specs.filter).toBe("filterh");
    });

    it("should match snapshot (no filter)", () => {
      const file = new fs.Dirent();
      file.name =
        "Light_LDN 1093_120.0s_Bin1_gain100_20240707-002348_-10.0C_0001.fit";
      file.path = "input/bar";
      const projectDirectory = "project/bar";
      const previousFile = null;

      const specs = getFileImageSpecFromFilename(
        file,
        projectDirectory,
        previousFile,
      );
      expect(specs).toMatchSnapshot();
      expect(specs.filter).toBeNull();
    });
  });

  it.each([
    { bulb: "120.0s", expected: 120000 },
    { bulb: "810.0ms", expected: 810 },
  ])("should parse bulb in ms (%s)", data => {
    const file = new fs.Dirent();
    file.name = `Light_LDN 1093_${data.bulb}_Bin1_gain100_20240707-002348_-10.0C_0001.fit`;
    file.path = "input/bar";
    const projectDirectory = "project/bar";
    const previousFile = null;

    const specs = getFileImageSpecFromFilename(
      file,
      projectDirectory,
      previousFile,
    );
    expect(specs.bulbMs).toBe(data.expected);
  });

  describe("getFitsFromDirectory", () => {
    it("should retrieve FITS files from a directory", () => {
      jest.spyOn(fs, "readdirSync").mockReturnValue([
        Object.assign(new fs.Dirent(), {
          name: "Light_LDN 1093_120.0s_Bin1_H_gain100_20240707-002348_-10.0C_0001.fit",
          isFile: () => true,
          path: "input/dir",
        }),
        Object.assign(new fs.Dirent(), {
          name: "Light_LDN 1093_120.0s_Bin1_H_gain100_20240707-002349_-10.0C_0002.fit",
          isFile: () => true,
          path: "input/dir",
        }),
        Object.assign(new fs.Dirent(), {
          name: "Light_LDN 1093_120.0s_Bin1_H_gain100_20240707-002348_-10.0C_0001_thn.jpg",
          isFile: () => true,
          path: "input/dir",
        }),
      ]);

      const result = getFitsFromDirectory({
        directory: "input/dir",
        projectDirectory: "project/dir",
      });

      expect(result).toHaveLength(2);
      expect(result[0].fileName).toBe(
        "Light_LDN 1093_120.0s_Bin1_H_gain100_20240707-002348_-10.0C_0001.fit",
      );
      expect(result[1].fileName).toBe(
        "Light_LDN 1093_120.0s_Bin1_H_gain100_20240707-002349_-10.0C_0002.fit",
      );
    });
  });

  describe("copyFileToProject", () => {
    it("should copy a file to the project directory", () => {
      const file = {
        projectDirectory: "project/dir",
        fileName: "file1.fit",
        sourceFilePath: "input/dir/file1.fit",
      } as FileImageSpec;

      jest.spyOn(fs, "existsSync").mockReturnValue(false);
      jest.spyOn(fs, "mkdirSync").mockImplementation(_ => "");
      jest.spyOn(fs, "copyFileSync").mockImplementation(() => {});

      copyFileToProject(file);

      expect(fs.existsSync).toHaveBeenCalledWith("project/dir");
      expect(fs.mkdirSync).toHaveBeenCalledWith("project/dir", {
        recursive: true,
      });
      expect(fs.copyFileSync).toHaveBeenCalledWith(
        "input/dir/file1.fit",
        "project/dir/file1.fit",
      );
    });
  });

  describe("matchSetFile", () => {
    it("should match Light with Dark", () => {
      const A = {
        type: "Light",
        bulb: "120.0s",
        bin: "Bin1",
        gain: 100,
      } as ImageSpec;
      const B = {
        type: "Dark",
        bulb: "120.0s",
        bin: "Bin1",
        gain: 100,
      } as ImageSpec;
      expect(matchSetFile(A, B)).toBe(true);
    });

    it("should match Light with Flat", () => {
      const A = { type: "Light", bin: "Bin1", filter: "H" } as ImageSpec;
      const B = { type: "Flat", bin: "Bin1", filter: "H" } as ImageSpec;
      expect(matchSetFile(A, B)).toBe(true);
    });

    it("should match Flat with Bias", () => {
      const A = { type: "Flat", bin: "Bin1", gain: 100 } as ImageSpec;
      const B = {
        type: "Bias",
        bin: "Bin1",
        gain: 100,
      } as ImageSpec;
      expect(matchSetFile(A, B)).toBe(true);
    });

    it("should not match Light with Bias", () => {
      const A = { type: "Light" } as ImageSpec;
      const B = { type: "Bias" } as ImageSpec;
      expect(matchSetFile(A, B)).toBe(false);
    });
  });

  describe("getImageSpecFromSetName", () => {
    it("should return the correct ImageSpec with filter", () => {
      const setName = "Flat_520.0ms_Bin1_H_gain0";
      const result = getImageSpecFromSetName(setName);
      expect(result).toEqual({
        setName,
        type: "Flat",
        bulb: "520.0ms",
        bin: "Bin1",
        filter: "H",
        gain: 0,
      });
    });

    it("should return the correct ImageSpec without filter", () => {
      const setName = "Flat_520.0ms_Bin1_gain0";
      const result = getImageSpecFromSetName(setName);
      expect(result).toEqual({
        setName,
        type: "Flat",
        bulb: "520.0ms",
        bin: "Bin1",
        filter: null,
        gain: 0,
      });
    });
  });
});
