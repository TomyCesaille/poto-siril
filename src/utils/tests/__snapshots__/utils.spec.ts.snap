// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getFileImageSpecFromFilename should compose a sequence, even if sequence not starting by 1 1`] = `
{
  "bin": "Bin1",
  "bulb": "120.0s",
  "bulbMs": 120000,
  "datetime": 2024-07-07T00:23:48.000Z,
  "extension": "fit",
  "fileName": "Light_LDN 1093_120.0s_Bin1_gain100_20240707-002348_-10.0C_0002.fit",
  "filter": null,
  "gain": 100,
  "projectFileDirectory": "project/bar/Light_120.0s_Bin1_gain100",
  "projectFilePath": "project/bar/Light_120.0s_Bin1_gain100/Light_LDN 1093_120.0s_Bin1_gain100_20240707-002348_-10.0C_0002.fit",
  "sequenceId": "20240707-002348",
  "sequencePosition": 2,
  "setName": "Light_120.0s_Bin1_gain100",
  "sourceFileDirectory": "input/bar",
  "sourceFilePath": "input/bar/Light_LDN 1093_120.0s_Bin1_gain100_20240707-002348_-10.0C_0002.fit",
  "temperature": -10,
  "type": "Light",
}
`;

exports[`utils getFileImageSpecFromFilename should compose a sequence, even when sequence is missing a file 1`] = `
{
  "bin": "Bin1",
  "bulb": "120.0s",
  "bulbMs": 120000,
  "datetime": 2024-07-07T00:23:48.000Z,
  "extension": "fit",
  "fileName": "Light_LDN 1093_120.0s_Bin1_gain100_20240707-002348_-10.0C_0003.fit",
  "filter": null,
  "gain": 100,
  "projectFileDirectory": "project/bar/Light_120.0s_Bin1_gain100",
  "projectFilePath": "project/bar/Light_120.0s_Bin1_gain100/Light_LDN 1093_120.0s_Bin1_gain100_20240707-002348_-10.0C_0003.fit",
  "sequenceId": "20240706-010203",
  "sequencePosition": 3,
  "setName": "Light_120.0s_Bin1_gain100",
  "sourceFileDirectory": "input/bar",
  "sourceFilePath": "input/bar/Light_LDN 1093_120.0s_Bin1_gain100_20240707-002348_-10.0C_0003.fit",
  "temperature": -10,
  "type": "Light",
}
`;

exports[`utils getFileImageSpecFromFilename should match snapshot (ASIAIR plan target format) 1`] = `
{
  "bin": "Bin1",
  "bulb": "120.0s",
  "bulbMs": 120000,
  "datetime": 2024-07-07T00:23:48.000Z,
  "extension": "fit",
  "fileName": "Light_LDN 1093_120.0s_Bin1_H_gain100_20240707-002348_-10.0C_0001.fit",
  "filter": "H",
  "gain": 100,
  "projectFileDirectory": "project/bar/H/Light_120.0s_Bin1_H_gain100",
  "projectFilePath": "project/bar/H/Light_120.0s_Bin1_H_gain100/Light_LDN 1093_120.0s_Bin1_H_gain100_20240707-002348_-10.0C_0001.fit",
  "sequenceId": "20240707-002348",
  "sequencePosition": 1,
  "setName": "Light_120.0s_Bin1_H_gain100",
  "sourceFileDirectory": "input/bar",
  "sourceFilePath": "input/bar/Light_LDN 1093_120.0s_Bin1_H_gain100_20240707-002348_-10.0C_0001.fit",
  "temperature": -10,
  "type": "Light",
}
`;

exports[`utils getFileImageSpecFromFilename should match snapshot (filter named with spaces) 1`] = `
{
  "bin": "Bin1",
  "bulb": "120.0s",
  "bulbMs": 120000,
  "datetime": 2024-07-07T00:23:48.000Z,
  "extension": "fit",
  "fileName": "Light_LDN 1093_120.0s_Bin1_filter h_gain100_20240707-002348_-10.0C_0001.fit",
  "filter": "filterh",
  "gain": 100,
  "projectFileDirectory": "project/bar/filterh/Light_120.0s_Bin1_filterh_gain100",
  "projectFilePath": "project/bar/filterh/Light_120.0s_Bin1_filterh_gain100/Light_LDN 1093_120.0s_Bin1_filter h_gain100_20240707-002348_-10.0C_0001.fit",
  "sequenceId": "20240707-002348",
  "sequencePosition": 1,
  "setName": "Light_120.0s_Bin1_filterh_gain100",
  "sourceFileDirectory": "input/bar",
  "sourceFilePath": "input/bar/Light_LDN 1093_120.0s_Bin1_filter h_gain100_20240707-002348_-10.0C_0001.fit",
  "temperature": -10,
  "type": "Light",
}
`;

exports[`utils getFileImageSpecFromFilename should match snapshot (no filter) 1`] = `
{
  "bin": "Bin1",
  "bulb": "120.0s",
  "bulbMs": 120000,
  "datetime": 2024-07-07T00:23:48.000Z,
  "extension": "fit",
  "fileName": "Light_LDN 1093_120.0s_Bin1_gain100_20240707-002348_-10.0C_0001.fit",
  "filter": null,
  "gain": 100,
  "projectFileDirectory": "project/bar/Light_120.0s_Bin1_gain100",
  "projectFilePath": "project/bar/Light_120.0s_Bin1_gain100/Light_LDN 1093_120.0s_Bin1_gain100_20240707-002348_-10.0C_0001.fit",
  "sequenceId": "20240707-002348",
  "sequencePosition": 1,
  "setName": "Light_120.0s_Bin1_gain100",
  "sourceFileDirectory": "input/bar",
  "sourceFilePath": "input/bar/Light_LDN 1093_120.0s_Bin1_gain100_20240707-002348_-10.0C_0001.fit",
  "temperature": -10,
  "type": "Light",
}
`;
